mm2src/mm2_metrics/src/recorder.rs

```diff
@@ -0,0 +1,211 @@
+}
+
+impl Default for MmRecorder {
+    fn default() -> Self {
+        Self {
+            registry: Registry::new(metrics_util::registry::GenerationalStorage::atomic()),
+        }
+    }
+}
+
+impl MmRecorder {
+    pub fn get_metrics(&self) -> Snapshot {
+        let counters = self
+            .registry
+            .get_counter_handles()
+            .into_iter()
+            .map(|(key, counter)| {
+                counter.get_generation();
+                let value = counter.get_inner().load(Ordering::Acquire); // This is a specific part for counter/gauge and histogram.
+                let inner = key_value_to_snapshot_entry(key.clone(), value);
+                (key.to_string(), inner)
+            })
+            .collect::<HashMap<String, HashMap<_, _>>>();
+
+        let gauges = self
+            .registry
+            .get_gauge_handles()
+            .into_iter()
+            .map(|(key, gauge)| {
+                gauge.get_generation();
+                let value = gauge.get_inner().load(Ordering::Acquire); // This is a specific part for counter/gauge and histogram.
+                let inner = key_value_to_snapshot_entry(key.clone(), value);
+                (key.to_string(), inner)
+            })
+            .collect::<HashMap<String, HashMap<_, _>>>();
+
+        let histograms = self
+            .registry
+            .get_histogram_handles()
+            .into_iter()
+            .map(|(key, histogram)| {
+                histogram.get_generation();
+                let value: f64 = histogram.get_inner().data().iter().sum(); // This is a specific part for counter/gauge and histogram.
+                let inner = key_value_to_snapshot_entry(key.clone(), value);
+                (key.to_string(), inner)
+            })
+            .collect::<HashMap<String, HashMap<_, _>>>();
+
+        Snapshot {
+            counters,
+            gauges,
+            histograms,
+        }
+    }
+
+    pub fn prepare_tag_metrics(&self) -> Vec<PreparedMetric> {
+        let mut output: Vec<PreparedMetric> = vec![];
+
+        for (key, counter) in self.registry.get_counter_handles() {
+            let value = counter.get_inner().load(Ordering::Acquire);
+            output.push(map_metric_to_prepare_metric(key, value));
+        }
+
+        for (key, gauge) in self.registry.get_gauge_handles() {
+            let value = gauge.get_inner().load(Ordering::Acquire);
+            output.push(map_metric_to_prepare_metric(key, value));
+        }
+
+        for (key, histogram) in self.registry.get_histogram_handles() {
+            let value: f64 = histogram.get_inner().data().iter().sum();
+            output.push(map_metric_to_prepare_metric(key, value as u64));
+        }
+
+        output
+    }
+
+    pub fn prepare_json(&self) -> MetricsJson {
+        let Snapshot {
+            counters,
+            gauges,
+            histograms,
+        } = self.get_metrics();
```

Comment:
Let's `->` means `used by`
As I can see,
`MmRecorder::get_metrics() -> MmRecorder::prepare_json() -> JsonObserver::observe() -> Metrics::collect_json()`.
The above functions are not used elsewhere.

Please also note that `MmRecorder::get_metrics()` and `MmRecorder::prepare_json()` iterate over all metrics both. It means that if we have three collections: `counters`, `gauges` and `histograms`, then these two functions will iterate 6 times to process them.
It's not efficient taking into account that we don't collect Json metrics into maps unlike `TagObserver`.

At the same time, such a chain of function calls makes it difficult to read the code. Such implementation could take place if `MmRecorder::get_metrics` or `MmRecorder::prepare_json()` are reused in other places, but they're not.

Please merge `MmRecorder::get_metrics` and `MmRecorder::prepare_json` into one function.